bokehjs/src/coffee/core/ui_events.coffee

```diff
@@ -61,36 +79,42 @@ export class UIEvents
       logger.debug("Button tool: #{type}")
       return
 
-    if et in ['pan', 'pinch', 'rotate']
-      logger.debug("Registering tool: #{type} for event '#{et}'")
-      if tool_view["_#{et}_start"]?
-        tool_view.listenTo(@, "#{et}:start:#{id}", tool_view["_#{et}_start"])
-      if tool_view["_#{et}"]?
-        tool_view.listenTo(@, "#{et}:#{id}",       tool_view["_#{et}"])
-      if tool_view["_#{et}_end"]?
-        tool_view.listenTo(@, "#{et}:end:#{id}",   tool_view["_#{et}_end"])
-    else if et == "move"
-      logger.debug("Registering tool: #{type} for event '#{et}'")
-      if tool_view._move_enter?
-        tool_view.listenTo(@, "move:enter:#{id}", tool_view._move_enter)
-      tool_view.listenTo(@, "move:#{id}", tool_view["_move"])
-      if tool_view._move_exit?
-        tool_view.listenTo(@, "move:exit:#{id}", tool_view._move_exit)
-    else
-      logger.debug("Registering tool: #{type} for event '#{et}'")
-      tool_view.listenTo(@, "#{et}:#{id}", tool_view["_#{et}"])
+    v = tool_view
+
+    switch et
+      when "pan"
+        if v._pan_start?    then v.connectTo(@pan_start,    (x) -> if x.id == id then v._pan_start(x.e))
+        if v._pan?          then v.connectTo(@pan,          (x) -> if x.id == id then v._pan(x.e))
+        if v._pan_end?      then v.connectTo(@pan_end,      (x) -> if x.id == id then v._pan_end(x.e))
+      when "pinch"
+        if v._pinch_start?  then v.connectTo(@pinch_start,  (x) -> if x.id == id then v._pinch_start(x.e))
+        if v._pinch?        then v.connectTo(@pinch,        (x) -> if x.id == id then v._pinch(x.e))
+        if v._pinch_end?    then v.connectTo(@pinch_end,    (x) -> if x.id == id then v._pinch_end(x.e))
+      when "rotate"
+        if v._rotate_start? then v.connectTo(@rotate_start, (x) -> if x.id == id then v._rotate_start(x.e))
+        if v._rotate?       then v.connectTo(@rotate,       (x) -> if x.id == id then v._rotate(x.e))
+        if v._rotate_end?   then v.connectTo(@rotate_end,   (x) -> if x.id == id then v._rotate_end(x.e))
+      when "move"
+        if v._move_enter?   then v.connectTo(@move_enter,   (x) -> if x.id == id then v._move_enter(x.e))
+        if v._move?         then v.connectTo(@move,         (x) -> if x.id == id then v._move(x.e))
+        if v._move_exit?    then v.connectTo(@move_exit,    (x) -> if x.id == id then v._move_exit(x.e))
+      when "tap"
+        if v._tap?          then v.connectTo(@tap,          (x) -> if x.id == id then v._tap(x.e))
+      when "press"
+        if v._press?        then v.connectTo(@press,        (x) -> if x.id == id then v._press(x.e))
+      when "scroll"
+        if v._scroll?       then v.connectTo(@scroll,       (x) -> if x.id == id then v._scroll(x.e))
+      else
```

Comment:
@canavandl, you may want to review changes to this file. What we have here isn't pretty, but events like `move:exit:#{id}` are a big no-go at this point.